File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items) {
            if (item.name.equals("Sulfuras, Hand of Ragnaros")) {
                continue;
            }
            if (item.name.equals("Aged Brie")) {
                increaseQuality(item);
            } else if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
                increaseQuality(item);
                if (item.sellIn < 11) {
                    increaseQuality(item);
                }
                if (item.sellIn < 6) {
                    increaseQuality(item);
                }
            } else {
                decreaseQuality(item);
            }

            item.sellIn = item.sellIn - 1;

            if (item.sellIn < 0) {
                if (item.name.equals("Aged Brie")) {
                    increaseQuality(item);
                } else {
                    if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
                        item.quality = item.quality - item.quality;
                    } else {
                        decreaseQuality(item);
                    }
                }
            }
        }
    }

    protected void decreaseQuality(Item item) {
        if (item.quality > 0) {
            item.quality = item.quality - 1;
        }
    }

    protected void increaseQuality(Item item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }
    }
}

File: php/src/GildedRose.php
<?php

declare(strict_types=1);

namespace GildedRose;

final class GildedRose
{
    /**
     * @var Item[]
     */
    public $items;

    private array $names = [
        'item1' => 'Aged Brie',
        'item2' => 'Backstage passes to a TAFKAL80ETC concert',
        'item3' => 'Sulfuras, Hand of Ragnaros'
    ];

    public function __construct(array $items)
    {
        $this->items = $items;
    }

    private function subDay($item)
    {
        $item->sell_in--;
    }

    private function isExpired($item)
    {
        return $item->sell_in < 0;
    }

    private function handleOutdated($item)
    {
        if ($item->name == $this->names['item1'] && $item->quality < 50) {
            $item->quality++;
        }

        if ($item->name == $this->names['item2']) {
            $item->quality = 0;
        }

        if (!in_array($item->name, $this->names) && $item->quality > 0) {
            $item->quality--;
        }
    }

    public function updateQuality(): void
    {
        foreach ($this->items as $item) {
            if ($item->name == $this->names['item1']) {
                $item->quality++;
            }

            if ($item->name == $this->names['item2']) {
                $item->quality++;
                if ($item->sell_in < 11) {
                    $item->quality++;
                }
                if ($item->sell_in < 6) {
                    $item->quality++;
                }
            }

            if ($item->name == $this->names['item3']) {
                continue;
            }

            if (!in_array($item->name, $this->names)) {
                $item->quality--;
            }

            if ($item->quality > 50) {
                $item->quality = 50;
            }

            if ($item->quality < 0) {
                $item->quality = 0;
            }

            $this->subDay($item);

            if ($this->isExpired($item)) {
                $this->handleOutdated($item);
            }
        }
    }

    public function getItem()
    {
        return $this->items;
    }
}
